### DISCUSSION TITLE
Milvus rapidly fills storage with a huge number of files

### DISCUSSION THREAD

**Discussion by drassokhin** [milvus-io/milvus#19774] [Q&A and General discussion]
I have been running experiments with fingerprint-based chemical similarity search and stumbled upon the issue where Milvus ran amok and filled the volume containing the docker container volumes with a huge number of files, apparently, when trying to compact the storage.

To run the experiment, I installed Milvus standalone by following these instructions: https://milvus.io/docs/v2.1.x/install_standalone-docker.md on a c5.4xlarge AWS instance with Red Hat Enterprise Linux 8 and 250GB volume mounted to /app, where /app/milvus/volumes directory was located.

I used the Molecule Similarity Search https://milvus.io/docs/v2.1.x/molecular_similarity_search.md as an example, but wrote my own implementation of the code that populated the Milvus database with data from ChEMBL release 31. I did not use MySQL to store the structures, instead, I defined the schema to include SMILES-encoded chemical structures as follows:

```python
VECTOR_DIMENSION = 1024

fld_idx = FieldSchema(name="id", dtype=DataType.INT64, descrition="int64 primary id",
                     is_primary=True, auto_id=True)
fld_ext_id = FieldSchema(name="ext_id", dtype=DataType.INT64, descrition="int64 external id")
fld_smiles = FieldSchema(name="smiles", dtype=DataType.VARCHAR, max_length=8000)
fld_fingerprint = FieldSchema(name="fingerprint", dtype=DataType.BINARY_VECTOR,
                              descrition="binary vector fingerprint",
                              dim=VECTOR_DIMENSION, is_primary=False)

schema = CollectionSchema(fields=[fld_idx, fld_ext_id, fld_smiles, fld_fingerprint],
                          description="SimSearch")
                          
collection = Collection(name=collection_name, schema=schema, using='default')
```

In this test, I did not create any similarity search indices.

While the code inserting the data into the collection in batches of 2000 records per ```collection.insert``` call was running, I monitored the number of files in ```/app/milvus/volumes``` with the following command:

```
find volumes -type f | wc -l
```

After 0.5M records were inserted, the number of files was 1,444, after 1M records it was 2,888. After 2M records were inserted, I stopped the loader (the entire ChEMBL 31 has 2,304,875 chemical structures). I expected the number of files in ```/app/milvus/volumes``` at that point to be about 6,000, but, to my surprise, it turned out to be much larger, 74,236, and then started to rapidly grow further, reaching 215,511 files in about one hour, while the database was idle, and no externally-initiated operations (searches, inserts, etc.) on it were being performed. After I left it running overnight in the idle mode, nearly entire 250GB /app volume was filled with an astronomical number of files, most of which located in ```/app/volumes/minio/.minio.sys/buckets/a-bucket/files/insert_log/```. At this point, I had to stop the container by running 

```
docker composer stop
```

and cleaning up ```/app/volumes``` by recursively deleting all files and subdirectories with the ```rm -r command```, which, due to a huge number of files and subdirectories in it, took about 15 minutes to run!

Questions: 1) Why did Milvus generate such a huge number of files in the buckets? 2) Did I do something wrong in terms of the installation, configuration, etc.? 3) How can I fix this issue, which prevents me from running any further tests?

**Comment by drassokhin** [comment]
UPDATE: I repeated the test and observed the same behavior. After the insertion of 1M records following the procedure I described above into an initially empty stand-alone Milvus instance, the number of files in volumes was 2832, and that number was not growing. After I inserted yet another 1M records (to the total of 2M), the number of files in volumes started growing rapidly and reached 136244, after which I stopped the containers. Most of those files are located in volumes/minio:

```
#-> docker compose start
[+] Running 3/3
 ⠿ Container milvus-minio       Started                                                                                        0.5s
 ⠿ Container milvus-etcd        Started                                                                                        0.6s
 ⠿ Container milvus-standalone  Started                                                                                        0.5s
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
2873
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
2873
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
3687
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
3715
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
96426
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
97486
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
103062
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116334
[root@AWSAFINVAL0004:/app/milvus]#
#-> docker compose stop
[+] Running 3/3
 ⠿ Container milvus-standalone  Stopped                                                                                        0.7s
 ⠿ Container milvus-minio       Stopped                                                                                        0.7s
 ⠿ Container milvus-etcd        Stopped                                                                                        0.2s
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116333
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116333
[root@AWSAFINVAL0004:/app/milvus]#
#-> docker compose start
[+] Running 3/3
 ⠿ Container milvus-etcd        Started                                                                                        0.4s
 ⠿ Container milvus-minio       Started                                                                                        0.5s
 ⠿ Container milvus-standalone  Started                                                                                        0.5s
[root@AWSAFINVAL0004:/app/milvus]#
#-> docker compose start
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116336
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116336
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116336
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116336
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
116336
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
129608
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes -type f | wc -l
136244
[root@AWSAFINVAL0004:/app/milvus]#
#-> docker compose stop
[+] Running 3/3
 ⠿ Container milvus-standalone  Stopped                                                                                        1.2s
 ⠿ Container milvus-minio       Stopped                                                                                        0.7s
 ⠿ Container milvus-etcd        Stopped                                                                                        0.2s
[root@AWSAFINVAL0004:/app/milvus]#
#->
#-> find volumes/etcd -type f | wc -l
3
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes/milvus -type f | wc -l
37
[root@AWSAFINVAL0004:/app/milvus]#
#-> find volumes/milvus -type f | wc -l
#-> find volumes/minio -type f | wc -l
136203
[root@AWSAFINVAL0004:/app/milvus]#

```
Does anyone have any idea what's going on? If I keep the containers up and running, in just a few hours the entire 250GB volume will be filled with files.

**Reply by drassokhin** [reply]
I have upgraded Milvus from 2.1.1 to 2.1.4, and the behavior has changed. Now it behaves as expected, that is, I no longer observe the runaway growth of the number of files in the docker volumes directory.

**Reply by yhmo** [reply]
Great! In v2.1.4 we did fix a couple of memory/storage issues.

**Reply by drassokhin** [reply]
@yhmo Correction: as it turned out, the update from 2.1.1. to 2.1.4 did not fix this bug (see also the discussion here: https://github.com/milvus-io/milvus/discussions/20322). However, yesterday I installed v.2.2 and repeated the same test with ChEMBL fingerprints. Happy to tell you that I could no longer observe the 'explosive number of files' behavior that actually prevented me from running any further tests with the previous versions. Also, the index progress info seems to return proper values. One minor difference I noticed is that, in previous versions, if a collection had no indices, the similarity search on it was executed by brute force (via full enumeration and distance calculation). In v. 2.2, if a collection has no index, a search fails with an exception. I still have no clue what caused this weird behavior in the previous versions of Milvus I tested, but hopefully I won't run into the same issue again after the latest update.

**Reply by drassokhin** [reply]
@yhmo Update: unfortunately, the update to v.2.2 did not fix this bug, I just had to wait longer to observe the explosive growth of the number of files.
```
#-> for d in *; do echo "$d"; find "$d" -type f | wc -l; done
etcd
7
milvus
60
minio
11611461
```

Now the entire 250GB volume is full (with the minio directory containing 11.6 million files!) and the Milvus service is not responding. The last line in the log file (which is also very large, about 2.5GB) is:
```
{"log":"[2022/12/01 11:36:18.417 +00:00] [WARN] [storage/minio_chunk_manager.go:188] [\"failed to put object\"] [path=files/insert_log/437714206848188420/437714206848188421/437740182544095050/1/437740182544095078] [error=\"Storage backend has reached its minimum free disk threshold. Please delete a few objects to proceed.\"]\n","stream":"stdout","time":"2022-12-01T11:36:18.417672546Z"}
```
I still have no idea what causes this odd behavior (probably, bugs in the db compacting implementation?) and how to get around it. Unfortunately, it prevents me from further evaluating the product :-(

**Reply by yhmo** [reply]
Yes, the compaction machinery might be a possible reason.  BTW: Is any special configuration for your minio service?

So, let's disable compaction in the milvus.yml.
You need to map the milvus.yml to outside the container, in the docker-compose.yml
```
standalone:
    container_name: milvus-standalone
    image: milvusdb/milvus:v2.2.0
    command: ["milvus", "run", "standalone"]
    environment:
      ETCD_ENDPOINTS: etcd:2379
      MINIO_ADDRESS: minio:9000
    volumes:
      - ${DOCKER_VOLUME_DIRECTORY:-.}/volumes/milvus:/var/lib/milvus
      - ${DOCKER_VOLUME_DIRECTORY:-.}/configs:/milvus/configs
```
You can copy the milvus.yml from the container(path= /milvus/configs/milvus.yml) to local path ${DOCKER_VOLUME_DIRECTORY:-.}/configs

Then set dataCoord.enableCompaction=false in the ${DOCKER_VOLUME_DIRECTORY:-.}/configs/milvus.yml
```
dataCoord:
  address: localhost
  port: 13333
  enableCompaction: false # Enable data segment compaction
```

And restart the milvus container.

**Reply by drassokhin** [reply]
@yhmo No, I have not changed the default configuration, with one minor exception: I remapped port 19530 to 1111 in docker-compose.yml:
```
 container_name: milvus-standalone
    image: milvusdb/milvus:v2.2.0
..........
    ports:
      - "1111:19530"
      - "9091:9091"
```
I had to do this due to some local security policies that prevented me from opening port 19530. However, I don't think this has anything to do with the behavior we are discussing. I will try to disable the compaction as you recommended and see whether this fixes the problem.

**Reply by drassokhin** [reply]
@yhmo When I disabled the compaction by remapping
```
- ${DOCKER_VOLUME_DIRECTORY:-.}/configs/milvus.yaml:/milvus/configs/milvus.yaml
```
and 
```
dataCoord:
  address: localhost
  port: 13333
  enableCompaction: false # Disable data segment compaction
```
in milvus.yaml, I get nil pointer dereference and segfault on startup, I am afraid I will have to give up :-( :

```
{"log":"[2022/12/02 22:31:40.853 +00:00] [INFO] [datacoord/garbage_collector.go:70] [\"GC with option\"] [enabled=true] [interval=1h0m0s] [missingTolerance=24h0m0s] [dropTolerance=24h0m0s]\n","stream":"stdout","time":"2022-12-02T22:31:40.853455142Z"}
{"log":"[2022/12/02 22:31:40.853 +00:00] [DEBUG] [datacoord/service.go:236] [\"DataCoord init done ...\"]\n","stream":"stdout","time":"2022-12-02T22:31:40.853483969Z"}
{"log":"panic: runtime error: invalid memory address or nil pointer dereference\n","stream":"stderr","time":"2022-12-02T22:31:40.856682322Z"}
{"log":"[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x2696c8b]\n","stream":"stderr","time":"2022-12-02T22:31:40.856689074Z"}
{"log":"\n","stream":"stderr","time":"2022-12-02T22:31:40.856692741Z"}
{"log":"goroutine 413 [running]:\n","stream":"stderr","time":"2022-12-02T22:31:40.856695639Z"}
{"log":"github.com/milvus-io/milvus/internal/datacoord.(*Server).Start(0xc0013beb60)\n","stream":"stderr","time":"2022-12-02T22:31:40.856698513Z"}
{"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:305 +0x2b\n","stream":"stderr","time":"2022-12-02T22:31:40.856701977Z"}
{"log":"github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).start(0xc000b2a000)\n","stream":"stderr","time":"2022-12-02T22:31:40.856705582Z"}
{"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:188 +0x45\n","stream":"stderr","time":"2022-12-02T22:31:40.85671803Z"}
{"log":"github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run(0x1?)\n","stream":"stderr","time":"2022-12-02T22:31:40.856727559Z"}
{"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:238 +0x4f\n","stream":"stderr","time":"2022-12-02T22:31:40.856731157Z"}
{"log":"github.com/milvus-io/milvus/cmd/components.(*DataCoord).Run(0x5689900?)\n","stream":"stderr","time":"2022-12-02T22:31:40.856745169Z"}
{"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/components/data_coord.go:49 +0x1d\n","stream":"stderr","time":"2022-12-02T22:31:40.856780783Z"}
{"log":"github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1()\n","stream":"stderr","time":"2022-12-02T22:31:40.856785759Z"}
{"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:102 +0x18e\n","stream":"stderr","time":"2022-12-02T22:31:40.856788874Z"}
{"log":"created by github.com/milvus-io/milvus/cmd/roles.runComponent[...]\n","stream":"stderr","time":"2022-12-02T22:31:40.856795777Z"}
{"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:85 +0x18a\n","stream":"stderr","time":"2022-12-02T22:31:40.856804502Z"}

```

**Reply by shieldlin** [reply]
same problem met

**Reply by yhmo** [reply]
We have made a fix to avoid fragmented files for binary vectors: https://github.com/milvus-io/milvus/pull/20971. I suppose it will be available in the new minor version 2.2.1(which will be released in a couple of days)

Recently, some other users report the similar problem(the data in minio rapidly fills storage), still under investigation.

**Reply by xiaofan-luan** [reply]
> invalid memory address or nil pointer de

There seems to be a panic 

> @yhmo When I disabled the compaction by remapping
> 
> ```
> - ${DOCKER_VOLUME_DIRECTORY:-.}/configs/milvus.yaml:/milvus/configs/milvus.yaml
> ```
> 
> and
> 
> ```
> dataCoord:
>   address: localhost
>   port: 13333
>   enableCompaction: false # Disable data segment compaction
> ```
> 
> in milvus.yaml, I get nil pointer dereference and segfault on startup, I am afraid I will have to give up :-( :
> 
> ```
> {"log":"[2022/12/02 22:31:40.853 +00:00] [INFO] [datacoord/garbage_collector.go:70] [\"GC with option\"] [enabled=true] [interval=1h0m0s] [missingTolerance=24h0m0s] [dropTolerance=24h0m0s]\n","stream":"stdout","time":"2022-12-02T22:31:40.853455142Z"}
> {"log":"[2022/12/02 22:31:40.853 +00:00] [DEBUG] [datacoord/service.go:236] [\"DataCoord init done ...\"]\n","stream":"stdout","time":"2022-12-02T22:31:40.853483969Z"}
> {"log":"panic: runtime error: invalid memory address or nil pointer dereference\n","stream":"stderr","time":"2022-12-02T22:31:40.856682322Z"}
> {"log":"[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x2696c8b]\n","stream":"stderr","time":"2022-12-02T22:31:40.856689074Z"}
> {"log":"\n","stream":"stderr","time":"2022-12-02T22:31:40.856692741Z"}
> {"log":"goroutine 413 [running]:\n","stream":"stderr","time":"2022-12-02T22:31:40.856695639Z"}
> {"log":"github.com/milvus-io/milvus/internal/datacoord.(*Server).Start(0xc0013beb60)\n","stream":"stderr","time":"2022-12-02T22:31:40.856698513Z"}
> {"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:305 +0x2b\n","stream":"stderr","time":"2022-12-02T22:31:40.856701977Z"}
> {"log":"github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).start(0xc000b2a000)\n","stream":"stderr","time":"2022-12-02T22:31:40.856705582Z"}
> {"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:188 +0x45\n","stream":"stderr","time":"2022-12-02T22:31:40.85671803Z"}
> {"log":"github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run(0x1?)\n","stream":"stderr","time":"2022-12-02T22:31:40.856727559Z"}
> {"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:238 +0x4f\n","stream":"stderr","time":"2022-12-02T22:31:40.856731157Z"}
> {"log":"github.com/milvus-io/milvus/cmd/components.(*DataCoord).Run(0x5689900?)\n","stream":"stderr","time":"2022-12-02T22:31:40.856745169Z"}
> {"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/components/data_coord.go:49 +0x1d\n","stream":"stderr","time":"2022-12-02T22:31:40.856780783Z"}
> {"log":"github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1()\n","stream":"stderr","time":"2022-12-02T22:31:40.856785759Z"}
> {"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:102 +0x18e\n","stream":"stderr","time":"2022-12-02T22:31:40.856788874Z"}
> {"log":"created by github.com/milvus-io/milvus/cmd/roles.runComponent[...]\n","stream":"stderr","time":"2022-12-02T22:31:40.856795777Z"}
> {"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:85 +0x18a\n","stream":"stderr","time":"2022-12-02T22:31:40.856804502Z"}
> ```

There seems to be a panic here? did you check why?

**Reply by yhmo** [reply]
> > invalid memory address or nil pointer de
> 
> There seems to be a panic
> 
> > @yhmo When I disabled the compaction by remapping
> > ```
> > - ${DOCKER_VOLUME_DIRECTORY:-.}/configs/milvus.yaml:/milvus/configs/milvus.yaml
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and
> > ```
> > dataCoord:
> >   address: localhost
> >   port: 13333
> >   enableCompaction: false # Disable data segment compaction
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in milvus.yaml, I get nil pointer dereference and segfault on startup, I am afraid I will have to give up :-( :
> > ```
> > {"log":"[2022/12/02 22:31:40.853 +00:00] [INFO] [datacoord/garbage_collector.go:70] [\"GC with option\"] [enabled=true] [interval=1h0m0s] [missingTolerance=24h0m0s] [dropTolerance=24h0m0s]\n","stream":"stdout","time":"2022-12-02T22:31:40.853455142Z"}
> > {"log":"[2022/12/02 22:31:40.853 +00:00] [DEBUG] [datacoord/service.go:236] [\"DataCoord init done ...\"]\n","stream":"stdout","time":"2022-12-02T22:31:40.853483969Z"}
> > {"log":"panic: runtime error: invalid memory address or nil pointer dereference\n","stream":"stderr","time":"2022-12-02T22:31:40.856682322Z"}
> > {"log":"[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x2696c8b]\n","stream":"stderr","time":"2022-12-02T22:31:40.856689074Z"}
> > {"log":"\n","stream":"stderr","time":"2022-12-02T22:31:40.856692741Z"}
> > {"log":"goroutine 413 [running]:\n","stream":"stderr","time":"2022-12-02T22:31:40.856695639Z"}
> > {"log":"github.com/milvus-io/milvus/internal/datacoord.(*Server).Start(0xc0013beb60)\n","stream":"stderr","time":"2022-12-02T22:31:40.856698513Z"}
> > {"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/datacoord/server.go:305 +0x2b\n","stream":"stderr","time":"2022-12-02T22:31:40.856701977Z"}
> > {"log":"github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).start(0xc000b2a000)\n","stream":"stderr","time":"2022-12-02T22:31:40.856705582Z"}
> > {"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:188 +0x45\n","stream":"stderr","time":"2022-12-02T22:31:40.85671803Z"}
> > {"log":"github.com/milvus-io/milvus/internal/distributed/datacoord.(*Server).Run(0x1?)\n","stream":"stderr","time":"2022-12-02T22:31:40.856727559Z"}
> > {"log":"\u0009/go/src/github.com/milvus-io/milvus/internal/distributed/datacoord/service.go:238 +0x4f\n","stream":"stderr","time":"2022-12-02T22:31:40.856731157Z"}
> > {"log":"github.com/milvus-io/milvus/cmd/components.(*DataCoord).Run(0x5689900?)\n","stream":"stderr","time":"2022-12-02T22:31:40.856745169Z"}
> > {"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/components/data_coord.go:49 +0x1d\n","stream":"stderr","time":"2022-12-02T22:31:40.856780783Z"}
> > {"log":"github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1()\n","stream":"stderr","time":"2022-12-02T22:31:40.856785759Z"}
> > {"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:102 +0x18e\n","stream":"stderr","time":"2022-12-02T22:31:40.856788874Z"}
> > {"log":"created by github.com/milvus-io/milvus/cmd/roles.runComponent[...]\n","stream":"stderr","time":"2022-12-02T22:31:40.856795777Z"}
> > {"log":"\u0009/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:85 +0x18a\n","stream":"stderr","time":"2022-12-02T22:31:40.856804502Z"}
> > ```
> 
> There seems to be a panic here? did you check why?

Just check. The crash bug was fixed by this commit: https://github.com/milvus-io/milvus/pull/20581/files
But this commit is not in v2.2, set `enableCompaction` to be false and start the server will always crash v2.2.

**Reply by yhmo** [reply]
@drassokhin 
Since a commit was not included in v2.2, we can't disable compaction.
To be honest, the root cause of storage rapidly growing is still unknown. The next minor version v2.2.1 make a fix for binary vectors to avoid fragmented files, but I am not sure it is the root cause. Now we are looking for more logs from users who encounter this problem. If the problem still reproduces on your machine, please collect the logs for use, it would be helpful.

**Reply by yhmo** [reply]
@drassokhin 
I believe we have found the root cause of this issue. There is a logic bug in the compaction machinery. Another user provided a full log when he encountered this problem, and finally we found the root cause. I create an issue to trace: https://github.com/milvus-io/milvus/issues/21094

We will make a fix for this bug in the next minor version 2.2.1.
